src/System.Data.Common/src/System/Data/Common/DbProviderFactory.netcoreapp.cs

```diff
@@ -0,0 +1,56 @@
+// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+// See the LICENSE file in the project root for more information.
+
+namespace System.Data.Common
+{
+    public abstract partial class DbProviderFactory
+    {
+        private bool? _canCreateDataAdapter;
+        private bool? _canCreateCommandBuilder;
+
+        public virtual bool CanCreateDataAdapter
+        {
+            get
+            {
+                if (!_canCreateDataAdapter.HasValue)
```

Comment:
Is this code thread safe? I actually cannot think of any race, but bringing it up in case others can. E.g. I am not sure if Nullable<T> assignment is atomic, but I am not sure it matters because worse case you may end up creating the adapter multiple times but I don't think the property will return wrong results.

Reply:
Assignment of a reference is atomic, i.e. no value tearing can occur. So yes, at worst multiple data adapters may get created, but I think the code is be safe.

Reply:
[Here's the spec paragraph on reference atomicity](https://docs.microsoft.com/en-us/dotnet/csharp/language-reference/language-specification/variables#atomicity-of-variable-references): 

> Reads and writes of the following data types are atomic: bool, char, byte, sbyte, short, ushort, uint, int, float, and reference types. In addition, reads and writes of enum types with an underlying type in the previous list are also atomic. Reads and writes of other types, including long, ulong, double, and decimal, as well as user-defined types, are not guaranteed to be atomic. Aside from the library functions designed for that purpose, there is no guarantee of atomic read-modify-write, such as in the case of increment or decrement.